compiler/rustc_infer/src/infer/higher_ranked/mod.rs

```diff
@@ -100,11 +103,12 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
             })
         };
 
-        let (result, map) = self.tcx.replace_bound_vars(binder, fld_r, fld_t, fld_c);
+        let result =
+            self.tcx.replace_escaping_bound_vars(binder.skip_binder(), fld_r, fld_t, fld_c);
 
         // If there were higher-ranked regions to replace, then actually create
         // the next universe (this avoids needlessly creating universes).
-        if !map.is_empty() {
```

Comment:
Wait...is this wrong?

I think this should be taking into account types and consts too...

Reply:
Do we see bound types in practice? Like other than in canonical queries, which I assume we liberate elsewhere? (maybe not)

Reply:
We don't :) At least not much today. Which is likely why this isn't doesn't fail everywhere. But in the future, we'll probably see more.

Reply:
Is the wrongness pre-existing? This change looks ok to me from a mechanical refactoring point of view, but I don't know much about this code in general.

Reply:
Yeah, looks preexisting.

Reply:
So do you have any concern with this PR being merged?

Reply:
Nope! But someone (probably me) should file an issue so this doesn't get forgotten.